### DISCUSSION TITLE
Is there a way to force static rendering when getServerSideProps is present?

### DISCUSSION THREAD

**Discussion by davidbarratt** [vercel/next.js#12987] [Help]
I want to have a page that is render staticlly and dynamiclly. The reason I want to do this is because I want to have the HTTP Cache offer a stale-while-revalidate strategy. 

If user vists `/[..slug]` I want to send them the default page (and the API request will be made on the client). The next visitor will get the cached version so they can skip the API request, but will again update the content after the request has been returned.

In addition, I do **not** want the client to call on routing `getServerSideProps()` over HTTP, I want it to fallback to it's client-side request.

To put it another way, I want to force static generation, but when requested through the server, I want it to get the props.

Is there a way to do this? I thought about using the `fallback` option on `getStaticPaths()` but I don't actually want it to generate new HTML pages, I just want it to run `getServerSideProps()` if that is present.

**Comment by davidbarratt** [comment]
Maybe I could have a custom [Document](https://nextjs.org/docs/advanced-features/custom-document) and do it that way? Maybe call a method of a different name?

**Comment by Timer** [comment]
Have you considered using `getStaticProps` with the new `unstable_revalidate` option that's in beta for 9.4?

https://nextjs.org/blog/next-9-4#incremental-static-regeneration-beta

**Reply by davidbarratt** [reply]
ooo! that is very close to what I want!

I guess what I would like is a "fallback" option of "ssr" (or something like that). that wouldn't generate HTML at all (I could potentially have **millions** of routes, so I'd rather not generate any HTML other than the "default").

**Reply by davidbarratt** [reply]
Basically I'd just have a single static route `/slug` (or whatever) and everything else would be SSR. Then I'd configure the HTTP Cache (Varnish or whatever) to do it's thing. Sometimes the user might get the page on `/slug` and sometimes they might get the prerendered page, but it's ok either way. :)

**Reply by Timer** [reply]
We're adding a fallback option that uses SSR, but we're only going to enable it for crawlers. Why don't you want to implement a HTML fallback version of your page?

**Reply by davidbarratt** [reply]
> Why don't you want to implement a HTML fallback version of your page?

1. I run the server in a docker container, so generating HTML pages would cause the container to no longer be stateless. Updating the container would likewise be difficult because the volume would need to be merged with the updated static routes.
2. One app has an _unlimited_ number of routes. I imagine will end up in the 10s of millions.
3. Another app uses a query string for routing (for better or worse), but has either today's date or a future date in the query string. Once we roll over midnight, none of the routes from today are useful and can be deleted.

All of these problems are solved by using an HTTP Cache like Varnish.

What I am looking for is something like this:
1. Static generation of a "default" route
2. Dynamic generation of all other routes (which is basically every route because nothing will inherently match the default).
3. Do not call the SSR function over HTTP when routing on the client

Another way to do this would be to add another page function... something like `getServerOnlyProps()` ? That function wouldn't prevent static generation, and wouldn't be called over HTTP.

Or... having a config option like `forceStaticGeneration` (default `false`) and `clientRoutingServerProps` (default `true`)

Or.... having a `fallback` that is "ssr" (or something like that).

**Reply by davidbarratt** [reply]
To put this at scale... 1 million routes at ~1mb each would result in about 1 terabyte (!) of data.

**Reply by davidbarratt** [reply]
Varnish will automatically evict the last used objects in order to keep within the defined quota. I don't think static generation is going to do that for me. :(

**Reply by davidbarratt** [reply]
I suppose a work around would be creating two pages (one static, one SSR) that render the same compoment? Though that will still prevent `next export` from running and wont prevent `getServerSideProps()` from being called over HTTP on the client. :(
